### ISSUE TITLE
Powershell `failOnStderr: false` not working

### ISSUE THREAD

**Issue by djee-ms** [microsoft/azure-pipelines-yaml#306]
I am trying to do a manual checkout of a second repository due to the lack of support (yet) of multi-repositories (see #103). I use a powershell task running a powershell script from a file, which executes `git` commands. Some of these commands output to `stderr`, as documented by git. This is interpreted by default by the powershell task as an error, and fails the task spuriously.

I tried to use the `failOnStderr` flag on that task and set it to `false`, but this doesn't seem to do anything. I also tried using `errorActionPreference` but I am skeptical about that because I do not want to avoid **all** errors in that script, just the spurious ones from writing to `stderr`.

I have another script using git from powershell which does not exhibit this issue. That other script is running on an Azure agent. The script which fails spuriously and for which `failOnStderr: false` is not working is running on a hosted agent on a VM I own. Is this an expected behavior? Are there some configuration steps to execute on my agent to make it behave like ADO ones? Or is it a bug?

_Note:_ At the moment I am working around that by wrapping each `git` command into a call to `Invoke-Expression`, capturing `stderr` and redirecting to `stdout`, and checking the return code for error. This seems to be what the built-in `checkout` task is doing too. I was not able to use PS redirection (_e.g._ `*>&1`); the interpreter always throws an exception.

**Comment by SQLHorizons** [comment]
@djee-ms 
Coincidence that I also do a manual clone with git through PowerShell and in a complex staged pipeline have received the same issue with the stage raising a failure due to standard out from git.  However when I create a simple test with basic PowerShell stages I cannot reproduce the issue.  This all started for me when I proceeded the git checkout with a call to a Jenkins pipeline.  My simple test:

```yaml

parameters:
  proxy:

steps:
  - powershell: |

      Try {

        if( -not( [string]::IsNullOrEmpty( "${{ parameters.proxy }}" ) )) {

          $env:HTTP_PROXY  = "${{ parameters.proxy }}"
          $env:HTTPS_PROXY = "${{ parameters.proxy }}"

        }

        ##  location of git exe.
        $git = "$(Agent.HomeDirectory)\externals\git\cmd\git.exe"

        ##  remove old repository.
        $release = @{
          Path        = "$(System.DefaultWorkingDirectory)\release"
          Recurse     = $true
          ErrorAction = "Ignore"
          Force       = $true
        }
        Write-Host "##vso[task.logIssue type=warning;] Delete directory: $($release.Path)"
        $null = Remove-Item @release

        if (Test-Path $($release.Path)) {
          Write-Host "##vso[task.logIssue type=warning;] Failed to delete directory: $($release.Path)"
        }
        else {
          Write-Host "##vso[task.logIssue type=warning;] Deleted directory: $($release.Path)"
        }

        ##  clone repository.
        $GitClone = @{
          FilePath              = $git
          ArgumentList          =  @(
            "clone"
            "https://$(xTokenAuth):$(token)@$(repository)",
            "--branch $(branch)",
            "release"
          )
          NoNewWindow           = $true
          Wait                  = $true
          PassThru              = $true
        }
        Write-Host "clone project repository: $(repository)."
        $null = Start-Process @GitClone

      }
      Catch [System.Exception] {

        exit 1

      }

    displayName: "clone project repository"
    workingDirectory: "$(System.DefaultWorkingDirectory)"
    errorActionPreference: stop
    failOnStderr: $(failOnStderr)
    continueOnError: false
    enabled: true
    timeoutInMinutes: 1

```
```yaml

name: __$(Date:yy.MM.dd).$(rev:rr)

trigger: none
pr: none

stages:
- stage: Start
  displayName: Start
  jobs:
  - job:
    pool:
      vmImage: windows-latest
    steps:
    - powershell: Write-Host ""
      failOnStderr: true

- stage: Test
  displayName: Functional Tests
  dependsOn: Start
  condition: succeeded()
  jobs:
  - deployment:
    displayName: git test
    timeoutInMinutes: 2
    pool:
      vmImage: windows-latest
    environment: sql-npe
    strategy:
      runOnce:
        deploy:
          steps:
          - template: ../.jobs/steps/step-cloneRepository.yml

- stage: End
  displayName: End
  dependsOn: Test
  condition: succeeded()
  jobs:
  - job:
    pool:
      vmImage: windows-latest
    steps:
    - powershell: Write-Host ""
      failOnStderr: true

```

**Comment by Sharparam** [comment]
We're experiencing this issue as well, using the PowerShell task to run a Cake script which in turn runs `npm install`. NPM writes some warnings/notices to STDERR, which incorrectly fails the build, and the PowerShell task seems to completely ignore the `failOnStderr: false` that we've set.

**Comment by fibann** [comment]
Any plan to get this fixed or at least investigated?

Recap of my experience:
- `failOnStderr: false` has no effect
- redirecting stderr with `2>&1` has no effect
- `-ErrorAction Ignore` has no effect
- the only thing that seems to work is `$ErrorActionPreference = "SilentlyContinue"` before the command (which is better than setting `errorActionPreference` for the whole script, but really verbose if you push/set/pop the value for every git command).

**Comment by sergiubologa** [comment]
I'm facing the exact same issue. Has anyone any updates on this?

**Comment by sergiubologa** [comment]
Solved my issue by upgrading from Powershell 1 to Powershell 2.

**Comment by SowmyaDantuluri** [comment]
Even for ps2. It is failing for me

Yaml:
task: PowerShell@2
  displayName: 'Running verisoning script'
  inputs:
    targetType: filePath
    filePath: ./Scripts/Versioning.ps1
    errorActionPreference: stop
    failOnStandardError: false
    ignoreLASTEXITCODE: false
    pwsh: false

Error: 
![image](https://user-images.githubusercontent.com/49900591/78239632-95f9a280-74fb-11ea-87b9-99d1babd075c.png)

**Comment by Saibamen** [comment]
Moved into Developer Community: https://developercommunity.visualstudio.com/content/problem/1007471/powershell-failonstderr-false-not-working-1.html
Please use "Follow" button on the right to track it

**Comment by Saibamen** [comment]
@djee-ms: Please go to this link and answer Microsoft team: https://developercommunity.visualstudio.com/content/problem/1007471/powershell-failonstderr-false-not-working-1.html

**Comment by ngcdev** [comment]
@Saibamen 
How do we reopen this issue? The link to the Developer Community is just closed with no solution. I have proof that this is not working with a very simple task (personal hosted VM):

```yaml
- pwsh: |
    Write-Error "Just some error message"
    Write-Output "This should display but it doesn't"
  failOnStderr: false
```
The above should print the "error" and since `failOnStderr` is set to false, it should continue. However, the task halts immediately and the second message is never displayed. The output is actually this:
```
Starting: PowerShell
==============================================================================
Task         : PowerShell
Description  : Run a PowerShell script on Linux, macOS, or Windows
Version      : 2.247.1
Author       : Microsoft Corporation
Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
==============================================================================
Generating script.
========================== Starting Command Output ===========================
/usr/bin/pwsh -NoLogo -NoProfile -NonInteractive -Command . '/home/buildagent/agent3/_work/_temp/6e84ea40-d7b4-42ec-a845-b3ca19eef8a8.ps1'
Write-Error: Just some error message

##[error]PowerShell exited with code '1'.
Finishing: PowerShell
```
Switching `failOnStderr` to true makes no difference.

Here's the full debug log of that simple task:
```
2025-02-26T21:31:19.5528902Z ##[debug]Evaluating condition for step: 'PowerShell'
2025-02-26T21:31:19.5529968Z ##[debug]Evaluating: SucceededNode()
2025-02-26T21:31:19.5530302Z ##[debug]Evaluating SucceededNode:
2025-02-26T21:31:19.5530778Z ##[debug]=> True
2025-02-26T21:31:19.5531168Z ##[debug]Result: True
2025-02-26T21:31:19.5531517Z ##[section]Starting: PowerShell
2025-02-26T21:31:19.5724610Z ==============================================================================
2025-02-26T21:31:19.5724769Z Task         : PowerShell
2025-02-26T21:31:19.5724847Z Description  : Run a PowerShell script on Linux, macOS, or Windows
2025-02-26T21:31:19.5724977Z Version      : 2.247.1
2025-02-26T21:31:19.5725053Z Author       : Microsoft Corporation
2025-02-26T21:31:19.5725378Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/powershell
2025-02-26T21:31:19.5725524Z ==============================================================================
2025-02-26T21:31:20.4036502Z ##[debug]Agent environment resources - Disk: C:\ Available 880158.00 MB out of 1665047.00 MB, Memory: Used 70744.00 MB out of 130976.00 MB, CPU: Usage 24.05%
2025-02-26T21:31:20.9352509Z ##[debug]VstsTaskSdk 0.21.0 commit 9abb170036d3b4cc1781def2efa436d9b059104c
2025-02-26T21:31:21.1833790Z ##[debug]Loading module from path 'C:\agents\agent0\_work\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.247.1\ps_modules\Sanitizer\Sanitizer.psm1'.
2025-02-26T21:31:21.2029444Z ##[debug]$OVERRIDING $global:DebugPreference from 'Continue' to 'SilentlyContinue'.
2025-02-26T21:31:21.2077491Z ##[debug]Loading resource strings from: C:\agents\agent0\_work\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.247.1\ps_modules\Sanitizer/module.json
2025-02-26T21:31:21.2195502Z ##[debug]Loaded 1 strings.
2025-02-26T21:31:21.2266994Z ##[debug]SYSTEM_CULTURE: 'en-US'
2025-02-26T21:31:21.2314846Z ##[debug]Loading resource strings from: C:\agents\agent0\_work\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.247.1\ps_modules\Sanitizer\Strings\resources.resjson\en-US\resources.resjson
2025-02-26T21:31:21.2401365Z ##[debug]Loaded 1 strings.
2025-02-26T21:31:21.2808812Z ##[debug]Feature flag AZP_75787_ENABLE_NEW_LOGIC state: False
2025-02-26T21:31:21.2869402Z ##[debug]Feature flag AZP_75787_ENABLE_NEW_LOGIC_LOG state: False
2025-02-26T21:31:21.2926350Z ##[debug]Feature flag AZP_75787_ENABLE_COLLECT state: True
2025-02-26T21:31:21.3524065Z ##[debug]Exporting function 'Get-SanitizerFeatureFlags'.
2025-02-26T21:31:21.3556868Z ##[debug]Exporting function 'Get-SanitizerCallStatus'.
2025-02-26T21:31:21.3589576Z ##[debug]Exporting function 'Get-SanitizerActivateStatus'.
2025-02-26T21:31:21.3622275Z ##[debug]Exporting function 'Protect-ScriptArguments'.
2025-02-26T21:31:21.3658120Z ##[debug]Importing function 'Get-SanitizerActivateStatus'.
2025-02-26T21:31:21.3689390Z ##[debug]Importing function 'Get-SanitizerCallStatus'.
2025-02-26T21:31:21.3720199Z ##[debug]Importing function 'Get-SanitizerFeatureFlags'.
2025-02-26T21:31:21.3751080Z ##[debug]Importing function 'Protect-ScriptArguments'.
2025-02-26T21:31:21.4018146Z ##[debug]Entering C:\agents\agent0\_work\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.247.1\powershell.ps1.
2025-02-26T21:31:21.4133546Z ##[debug]Loading resource strings from: C:\agents\agent0\_work\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.247.1\task.json
2025-02-26T21:31:21.4553414Z ##[debug]Loaded 14 strings.
2025-02-26T21:31:21.4602700Z ##[debug]SYSTEM_CULTURE: 'en-US'
2025-02-26T21:31:21.4643607Z ##[debug]Loading resource strings from: C:\agents\agent0\_work\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.247.1\Strings\resources.resjson\en-US\resources.resjson
2025-02-26T21:31:21.4995952Z ##[debug]Loaded 14 strings.
2025-02-26T21:31:21.5311676Z ##[debug]INPUT_ERRORACTIONPREFERENCE: 'stop'
2025-02-26T21:31:21.5394046Z ##[debug]INPUT_WARNINGPREFERENCE: 'default'
2025-02-26T21:31:21.5439719Z ##[debug]INPUT_INFORMATIONPREFERENCE: 'default'
2025-02-26T21:31:21.5486475Z ##[debug]INPUT_VERBOSEPREFERENCE: 'default'
2025-02-26T21:31:21.5531431Z ##[debug]INPUT_DEBUGPREFERENCE: 'default'
2025-02-26T21:31:21.5576406Z ##[debug]INPUT_PROGRESSPREFERENCE: 'silentlyContinue'
2025-02-26T21:31:21.5619582Z ##[debug]INPUT_SHOWWARNINGS: 'false'
2025-02-26T21:31:21.5664579Z ##[debug] Converted to bool: False
2025-02-26T21:31:21.5710808Z ##[debug]INPUT_FAILONSTDERR: 'false'
2025-02-26T21:31:21.5746429Z ##[debug] Converted to bool: False
2025-02-26T21:31:21.5791609Z ##[debug]INPUT_IGNORELASTEXITCODE: 'false'
2025-02-26T21:31:21.5825285Z ##[debug] Converted to bool: False
2025-02-26T21:31:21.5871865Z ##[debug]INPUT_PWSH: 'true'
2025-02-26T21:31:21.5904825Z ##[debug] Converted to bool: True
2025-02-26T21:31:21.5948477Z ##[debug]INPUT_WORKINGDIRECTORY: 'C:\agents\agent0\_work\1546\s'
2025-02-26T21:31:21.6125692Z ##[debug]Asserting container path exists: 'C:\agents\agent0\_work\1546\s'
2025-02-26T21:31:21.6174201Z ##[debug]INPUT_TARGETTYPE: 'inline'
2025-02-26T21:31:21.6230381Z ##[debug]INPUT_SCRIPT: 'Write-Error "Just some error message"
2025-02-26T21:31:21.6250736Z ##[debug]Write-Output "This should display but it doesn't"'
2025-02-26T21:31:21.6292104Z ##[debug]INPUT_RUNSCRIPTINSEPARATESCOPE: 'false'
2025-02-26T21:31:21.6323465Z ##[debug] Converted to bool: False
2025-02-26T21:31:21.6372628Z Generating script.
2025-02-26T21:31:21.6596202Z ##[debug]AGENT_VERSION: '3.234.0'
2025-02-26T21:31:21.6700068Z ##[debug]AGENT_TEMPDIRECTORY: 'C:\agents\agent0\_work\_temp'
2025-02-26T21:31:21.6734505Z ##[debug]Asserting container path exists: 'C:\agents\agent0\_work\_temp'
2025-02-26T21:31:21.6997540Z ##[debug]Asserting leaf path exists: 'C:\Program Files\PowerShell\7\pwsh.exe'
2025-02-26T21:31:21.7016353Z ========================== Starting Command Output ===========================
2025-02-26T21:31:21.7216270Z ##[debug]Entering Invoke-VstsTool.
2025-02-26T21:31:21.7333114Z ##[debug] Arguments: '-NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\agents\agent0\_work\_temp\46767385-43f6-46f5-9383-58f054dde425.ps1'"'
2025-02-26T21:31:21.7364131Z ##[debug] FileName: 'C:\Program Files\PowerShell\7\pwsh.exe'
2025-02-26T21:31:21.7395171Z ##[debug] WorkingDirectory: 'C:\agents\agent0\_work\1546\s'
2025-02-26T21:31:21.7462288Z ##[command]"C:\Program Files\PowerShell\7\pwsh.exe" -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command ". 'C:\agents\agent0\_work\_temp\46767385-43f6-46f5-9383-58f054dde425.ps1'"
2025-02-26T21:31:22.4316233Z C:\agents\agent0\_work\_temp\46767385-43f6-46f5-9383-58f054dde425.ps1 : [31;1mJust some error message
2025-02-26T21:31:22.4316602Z At line:1 char:1
2025-02-26T21:31:22.4317287Z + . 'C:\agents\agent0\_work\_temp\46767385-43f6-46f5-9383-58f054dde425. …
2025-02-26T21:31:22.4317642Z + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-02-26T21:31:22.4317920Z + CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
2025-02-26T21:31:22.4318243Z + FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException,46767385-43f6-46f5-9383-58f054dde425.ps1[0m
2025-02-26T21:31:22.5198535Z ##[debug]Exit code: 1
2025-02-26T21:31:22.5261389Z ##[debug]Leaving Invoke-VstsTool.
2025-02-26T21:31:22.5636094Z ##[error]PowerShell exited with code '1'.
2025-02-26T21:31:22.5644145Z ##[debug]Processed: ##vso[task.logissue type=error;source=TaskInternal]PowerShell exited with code '1'.
2025-02-26T21:31:22.5767799Z ##[debug]Processed: ##vso[task.complete result=Failed]Error detected
2025-02-26T21:31:22.5796107Z ##[debug]Leaving C:\agents\agent0\_work\_tasks\PowerShell_e213ff0f-5d5c-4791-802d-52ea3e7be1f1\2.247.1\powershell.ps1.
2025-02-26T21:31:22.6161511Z ##[section]Finishing: PowerShell


```

**Comment by Saibamen** [comment]
Create new issue ticket here https://developercommunity.visualstudio.com/content/problem/
